modules/core/src/main/scala/org/tessellation/infrastructure/rewards/Rewards.scala

```diff
@@ -0,0 +1,197 @@
+package org.tessellation.infrastructure.rewards
+
+import cats.MonadThrow
+import cats.data._
+import cats.effect.Async
+import cats.effect.std.Random
+import cats.syntax.applicative._
+import cats.syntax.applicativeError._
+import cats.syntax.contravariantSemigroupal._
+import cats.syntax.either._
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+import cats.syntax.semigroup._
+import cats.syntax.traverse._
+
+import scala.collection.immutable.{SortedMap, SortedSet}
+import scala.concurrent.duration.FiniteDuration
+import scala.math.Integral.Implicits._
+
+import org.tessellation.config.types.RewardsConfig
+import org.tessellation.dag.snapshot.SnapshotOrdinal
+import org.tessellation.domain.rewards._
+import org.tessellation.schema.ID.Id
+import org.tessellation.schema.address.Address
+import org.tessellation.schema.balance.Amount
+import org.tessellation.schema.transaction.{RewardTransaction, TransactionAmount}
+import org.tessellation.security.SecurityProvider
+import org.tessellation.security.hash.Hash
+
+import eu.timepit.refined.auto._
+import eu.timepit.refined.types.numeric._
+import io.estatico.newtype.ops.toCoercibleIdOps
+import org.typelevel.log4cats.slf4j.Slf4jLogger
+
+object Rewards {
+
+  def make[F[_]: Async: SecurityProvider](
+    config: RewardsConfig,
+    timeTriggerInterval: FiniteDuration,
+    softStaking: SoftStaking,
+    dtm: DTM,
+    stardustCollective: StardustCollective
+  ): Rewards[F] =
+    new Rewards[F] {
+
+      implicit val logger = Slf4jLogger.getLogger[F]
+
+      def calculateRewards(
+        snapshotHash: Hash,
+        snapshotOrdinal: SnapshotOrdinal,
+        facilitators: NonEmptySet[Id]
+      ): F[(Map[Address, Amount], SortedSet[RewardTransaction])] =
+        tryCalculateRewards(snapshotHash, snapshotOrdinal, timeTriggerInterval, facilitators).rethrowT
+
+      private[rewards] def tryCalculateRewards(
+        snapshotHash: Hash,
+        snapshotOrdinal: SnapshotOrdinal,
+        timeTriggerInterval: FiniteDuration,
+        facilitators: NonEmptySet[Id]
+      ): EitherT[F, RewardsError, (Map[Address, Amount], SortedSet[RewardTransaction])] =
+        EitherT
+          .fromEither(getRewardsPerSnapshot(snapshotOrdinal, timeTriggerInterval))
+          .biflatMap(
+            {
+              case LastEpochExceeded =>
+                EitherT.liftF {
+                  logger.warn("Couldn't distribute the rewards. Last epoch exceeded.") >>
```

Comment:
if last epoch exceeded, there should be no rewards I think 🤔

Reply:
And there are no rewards. We just return empty `Set`:

```scala
              case LastEpochExceeded =>
                EitherT.liftF {
                  logger.warn("Couldn't distribute the rewards. Last epoch exceeded.") >>
                    (Map.empty[Address, Amount], SortedSet.empty[RewardTransaction]).pure[F]
                }
```

I can change the warning to be more descriptive e.g. "Last epoch exceeded. Won't mint rewards" or sth